### PR TITLE
Constrained decoding

### PR DESCRIPTION
**Pull request by VictorOdede**
Implemented regex decoding for #30. This currently only works for HuggingFace transformers.

- Added a custom logits processor for masking next-sequence tokens
- implemented masking function to filter legal tokens
- Modified generate function to accept regex string and bias parameters
- Added custom stopping logic for debugging 
- Added a testing for output generation

### CODE REVIEW

**Comment by VictorOdede** [comment]
Hi @mmirman  @cartazio , I managed to get the regex querying working using a custom logits processor from Huggingface. Looking forward to your feedback .

**Comment by cartazio** [comment]
we fat fingured merging this, now reverted, can you create a fresh PR so we can review this?

**Comment by mmirman** [comment]
@VictorOdede Did you read the comments from https://github.com/anarchy-ai/LLM-VM/pull/100? As per the reLLM project we have been discussing how to widen the scope of this so as not to be redundant with that (and just be a wrapper over what they have).  It think it would be a good idea to get on a call with @itsmeashutosh43 in discord so we can figure out what to do about splitting up this work and keeping things consistent

**Comment by itsmeashutosh43** [comment]
Yes, I am on discord as Ash, feel free to ping me.

File: src/llm_vm/onsite_llm.py [hunk 1]
```diff
@@ -184,6 +190,190 @@ def asynctune():
...
         finetune()()
 
+
+class TokenStreamerAsStoppingCriterion:
+    def __init__(self, token_streamer):
+        self.token_streamer = token_streamer
+
+    def __call__(self, input_ids, scores, **kwargs):
+        if self.token_streamer is None:
+            return False
```
Comment (cartazio) [inline src/llm_vm/onsite_llm.py:200]: I think itd be appropriate to return None here

**Comment by VictorOdede** [comment]
> we fat fingured merging this, now reverted, can you create a fresh PR so we can review this?

I've opened a new PR for this

**Comment by VictorOdede** [comment]
> @VictorOdede Did you read the comments from #100? As per the reLLM project we have been discussing how to widen the scope of this so as not to be redundant with that (and just be a wrapper over what they have). It think it would be a good idea to get on a call with @itsmeashutosh43 in discord so we can figure out what to do about splitting up this work and keeping things consistent

Just read the conversation on #100 and I think that wrapping reLLM is a great solution but might be a bit restrictive if we want to add custom behaviour. The HF logit processor allows more control over token predictions and biases for each generation. It also allows us to use the generate function which is a familiar interface for most developers.  I'm open to discussing this further over a call. Can we set up a chat tomorrow? @itsmeashutosh43

File: src/llm_vm/onsite_llm.py (hunk 1 above)
Comment (VictorOdede) [reply src/llm_vm/onsite_llm.py:200]: I agree
